### PR TITLE
Adding modules for segmentation as performed in cell profiler

### PR DESCRIPTION
**Pull request by scottberry**
I developed these a while ago based on the CellProfiler segmentation modules but they seem to not be in the main repository. They work in my hands but I can't remember if there are some additional python dependencies that need to be installed. The only issue here is licensing, which I am not sure about. I just copied in the BSD 3-Clause License I found on the [CellProfiler code](https://github.com/CellProfiler/CellProfiler)

### CODE REVIEW

**Review by jluethi — COMMENTED** [review]
Would be great to have those modules! And the wrapping looks good. I have no idea whether that's ok license-wise. Do you know @bebosudo ?

File: jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+def filter_labels(primary_label_image, secondary_label_image):
+
+    labels_in = np.unique(primary_label_image) - [0]
+    labels_out = np.unique(secondary_label_image) - [0]
+
+    extra_labels = set(labels_out) - set(labels_in)
+    missing_labels = set(labels_in) - set(labels_out)
+
+    if len(extra_labels) != 0:
```
Comment (bebosudo) [inline jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py]:
```suggestion
    if extra_labels:
```
Standard data structures (such as sets), support boolean evaluation.
https://docs.python.org/3/library/stdtypes.html#truth-value-testing

File: jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+        logger.warn(
+            'Removing labels {} detected in secondary_label_image, which'
+            'are not present in primary_label_image'.format(extra_labels))
+        for label in extra_labels:
+            secondary_label_image[secondary_label_image == label] = 0
+    elif len(missing_labels) != 0:
```
Comment (bebosudo) [inline jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py]:
```suggestion
    elif missing_labels:
```

Comment (bebosudo) [reply jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py]: Are you sure elif is correct? If there were extra_labels, this second branch won't be executed.

File: jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    Analysis and Machine Intelligence, Vol. 13, No. 6, 583-598
+    '''
+
+    if not np.any(primary_label_image == 0):
+        secondary_label_image = primary_label_image
+    else:
+        n_objects = len(np.unique(primary_label_image[1:]))
+        logger.info(
+            'primary label image has %d objects',
+            n_objects - 1
```
Comment (bebosudo) [inline jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py]:
```suggestion
            'primary label image has {} objects'.format(n_objects - 1)
```
% is being deprecated.
Is there a reason why we reducing the number of objects by 1?

File: jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+    for i in range(1,np.max(secondary_label_image)):
+      borders_with_background = borders_with_background + mh.labeled.border(labeled=secondary_label_image,i=i,j=0)
+
+    all_borders = mh.labeled.borders(secondary_label_image)
+    touching_pixels = np.logical_xor(borders_with_background,all_borders)
+    secondary_label_image[touching_pixels!=0] = 0
+
+    n_objects = len(np.unique(secondary_label_image)[1:])
+    logger.info('identified %d objects', n_objects)
```
Comment (bebosudo) [inline jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py]:
```suggestion
    logger.info('identified {} objects'.format(n_objects))
```

File: jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+            secondary_label_image = np.zeros(
+                primary_label_image.shape, dtype=np.int32
+            )
+
+    # Tissuemaps label images must be separated by lines,
+    # need to identify the boundary pixels
+    borders_with_background = np.zeros(primary_label_image.shape, dtype=np.int32)
+
+    for i in range(1,np.max(secondary_label_image)):
+      borders_with_background = borders_with_background + mh.labeled.border(labeled=secondary_label_image,i=i,j=0)
```
Comment (bebosudo) [inline jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py]:
```suggestion
      borders_with_background += mh.labeled.border(labeled=secondary_label_image, i=i, j=0)
```

File: jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+    all_borders = mh.labeled.borders(secondary_label_image)
+    touching_pixels = np.logical_xor(borders_with_background,all_borders)
+    secondary_label_image[touching_pixels!=0] = 0
```
Comment (bebosudo) [inline jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py]:
```suggestion
    secondary_label_image[touching_pixels != 0] = 0
```

**Review by bebosudo — CHANGES_REQUESTED** [review]

**Comment by bebosudo** [comment]
License-wise I've found the following discussion:
https://opensource.stackexchange.com/questions/8107/distributing-bsd-3-code-under-the-apache-license-v2-only

I think the double license banner that Scott used is fine.
@scottberry: if you could link to the precise original commit you copied the code from it could be great.

The only missing bit would be to mention the fact that part of the code is double-licensed in the sub-project README:
https://github.com/pelkmanslab/TissueMAPS/blob/master/jtlibrary/LICENSE.txt

If you could add a couple of lines in a LICENSE.txt file in the main repo, indicating that there are several projects in this repo, and to refer to the different licenses, it would be great!

File: jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py (hunk 3 above)
Comment (scottberry) [reply jtlibrary/python/jtmodules/src/jtmodules/segment_secondary_cp.py]: Yes, 0 is not an object

**Review by bebosudo — APPROVED** [review]

**Comment by scottberry** [comment]
@bebosudo would you be able to add the licence info to jtlibrary README.txt for me?

**Comment by bebosudo** [comment]
Sure, I'll do.

**Pull request merged**